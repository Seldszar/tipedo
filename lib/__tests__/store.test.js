const Layer = require("../layer");
const Store = require("../store");

describe("Store", () => {
  let store;

  beforeEach(() => {
    store = new Store([
      new Layer("lorem", {
        lorem: "from lorem",
      }),
      new Layer("ipsum", {
        lorem: "from ipsum",
        ipsum: "from ipsum",
      }),
    ]);
  });

  test("should return an empty store", () => {
    expect(new Store()).toMatchSnapshot();
  });

  test("should return the `lorem` node", () => {
    expect(store.get("lorem")).toBe("from lorem");
  });

  test("should return `true` if a node exists", () => {
    expect(store.has("lorem")).toBe(true);
  });

  test("should return `false` if a node doesn't exist", () => {
    expect(store.has("amet")).toBe(false);
  });

  test("should include the nodes from the `lorem` layer", () => {
    const layer = store.layer("ipsum");
    const callback = jest.fn();

    store.on("change", callback);
    layer.clear();

    expect(callback).toHaveBeenCalled();
    expect([...store.keys()]).toMatchSnapshot();
  });

  test("should not emit a `change` event if a lower layer updates the same node as the higher one", () => {
    const layer = store.layer("ipsum");
    const callback = jest.fn();

    store.on("change", callback);
    layer.set("lorem", "from ipsum");

    expect(callback).not.toHaveBeenCalled();
    expect(store.get("lorem")).toBe("from lorem");
  });

  test("should emit a `change` event if a higher layer updates the same node as the lower one", () => {
    const layer = store.layer("lorem");
    const callback = jest.fn();

    store.on("change", callback);
    layer.set("ipsum", "from lorem");

    expect(callback).toHaveBeenCalled();
    expect(store.get("ipsum")).toBe("from lorem");
  });

  test("should walk through the store and return the nodes", () => {
    const callback = jest.fn();

    store.forEach(callback);

    expect(callback).toHaveBeenCalled();
    expect(callback.mock.calls).toMatchSnapshot();
  });

  test("should return the keys", () => {
    expect([...store.keys()]).toMatchSnapshot();
  });

  test("should return the values", () => {
    expect([...store.values()]).toMatchSnapshot();
  });

  test("should return the entries", () => {
    expect([...store.entries()]).toMatchSnapshot();
  });

  test("should return the JSON compatible object", () => {
    expect(store.toJSON()).toMatchSnapshot();
  });

  test("should return the identity", () => {
    expect(store.toString()).toMatchSnapshot();
  });

  test("should return an iterator", () => {
    const iterator = store[Symbol.iterator]();

    expect(iterator).toBeInstanceOf(Object);
    expect(iterator.next).toBeInstanceOf(Function);
  });
});
