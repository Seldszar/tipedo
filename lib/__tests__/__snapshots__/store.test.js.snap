// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Store should include the nodes from the \`lorem\` layer 1`] = `
Array [
  "lorem",
]
`;

exports[`Store should return an empty store 1`] = `
Object {
  "nodes": Object {},
}
`;

exports[`Store should return the JSON compatible object 1`] = `
Object {
  "nodes": Object {
    "ipsum": "from ipsum",
    "lorem": "from lorem",
  },
}
`;

exports[`Store should return the entries 1`] = `
Array [
  Array [
    "lorem",
    "from lorem",
  ],
  Array [
    "ipsum",
    "from ipsum",
  ],
]
`;

exports[`Store should return the identity 1`] = `"[object Store]"`;

exports[`Store should return the keys 1`] = `
Array [
  "lorem",
  "ipsum",
]
`;

exports[`Store should return the values 1`] = `
Array [
  "from lorem",
  "from ipsum",
]
`;

exports[`Store should walk through the store and return the nodes 1`] = `
Array [
  Array [
    "from lorem",
    "lorem",
  ],
  Array [
    "from ipsum",
    "ipsum",
  ],
]
`;
